jfsd/utils.py

```diff
@@ -10,28 +10,295 @@
 from jax.typing import ArrayLike
 
 from jfsd import ewald_tables, shear, thermal
-from jfsd import jaxmd_partition as partition
 from jfsd import jaxmd_space as space
 
 # Define types for the functions
 DisplacementFn = Callable[[Any, Any], Any]
 
+def preprocess_sparse_triangular(m_sparse, num_particles, max_nonzero_per_row):
+    """
+    Efficiently precomputes sparse lookup structures for forward substitution.
 
-@jit
-def chol_fac(matrix: ArrayLike) -> Array:
-    """Perform a Cholesky factorization of the input matrix.
+    Parameters:
+    - l_low: JAX sparse BCOO lower triangular matrix
+    - num_particles: int, number of particles
+
+    Returns:
+    - row_indices: (num_particles, max_nonzero_per_row) indices of nonzero entries
+    - row_values: (num_particles, max_nonzero_per_row) corresponding values
```

Comment:
Perhaps you can rewrite this to numpy docstyle so that it can be read by sphinx?